File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char,int> stringT;

        for(char c : t){
            stringT[c]++;
        }

        int have = 0;
        int need = stringT.size();
        int minLen = INT_MAX;
        int left = 0;
        int minStart = 0;

        unordered_map<char,int> window;

        for(int right = 0; right < s.length(); right++){
            char c = s[right];
            window[c]++;

            if(stringT.count(c) && window[c] == stringT[c]){
                have++;
            }

            while(have == need){
                if(right - left + 1 < minLen){
                    minLen = right - left + 1;
                    minStart = left;
                }

                char remove = s[left];
                window[remove]--;
                if(stringT.count(remove) && window[remove] < stringT[remove]){
                    have--;
                }

                left++;
            }
        }

        return minLen == INT_MAX ? "" : s.substr(minStart, minLen);
    }
};

File: 76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        stringT = defaultdict(int)

        for i in range(len(t)):
            stringT[t[i]] += 1
        
        have = 0
        need = len(stringT)
        minLen = float('inf')
        left = 0
        minstart = 0

        window = defaultdict(int)

        for right in range(len(s)):
            c = s[right]
            window[c] += 1

            if c in stringT and stringT[c] == window[c]:
                have += 1
            
            while have == need:
                if minLen > (right - left + 1):
                    minLen = (right - left + 1)
                    minstart = left
                    
                remove = s[left]
                window[remove] -= 1
                if remove in stringT and window[remove] < stringT[remove]:
                    have -= 1
                left += 1
        
        return "" if minLen == float('inf') else s[minstart: minLen+minstart]